File: cobalt/backend/domain/repositories/roles.py
#  Copyright (C) 2026 ArtoriasCode
#  Author: ArtoriasCode
#  Repository: https://github.com/ArtoriasCode/cobalt
#  SPDX-License-Identifier: AGPL-3.0-or-later

from abc import ABC, abstractmethod
from typing import Optional, List, Any

from domain.entities import (
    RoleEntity,
    RolesPageEntity,
    RolesGetPageEntity,
    RoleCreateEntity,
    RoleUpdateEntity
)


class AbstractRolesRepository(ABC):
    """
    Abstract roles repository.
    """

    @abstractmethod
    async def get_page(
        self,
        entity: RolesGetPageEntity,
        session: Optional[Any] = None
    ) -> RolesPageEntity:
        """
        Gets a paginated list of roles.

        Parameters:
        - entity: RolesGetPageEntity object.
        - session: Session object.

        Returns:
        - RolesPageEntity: RolesPageEntity object.
        """
        ...

    @abstractmethod
    async def get_one_by_id(
        self,
        role_id: int,
        session: Optional[Any] = None
    ) -> Optional[RoleEntity]:
        """
        Gets an existing role by ID.

        Parameters:
        - role_id: Role ID.
        - session: Session object.

        Returns:
        - RoleEntity: RoleEntity object.
        """
        ...

    @abstractmethod
    async def create_one(
        self,
        entity: RoleCreateEntity,
        session: Optional[Any] = None
    ) -> RoleEntity:
        """
        Creates a new role.

        Parameters:
        - entity: RoleCreateEntity object.
        - session: Session object.

        Returns:
        - RoleEntity: RoleEntity object.
        """
        ...

    @abstractmethod
    async def update_one(
        self,
        entity: RoleUpdateEntity,
        session: Optional[Any] = None
    ) -> Optional[RoleEntity]:
        """
        Updates an existing role.

        Parameters:
        - entity: RoleUpdateEntity object.
        - session: Session object.

        Returns:
        - RoleEntity: RoleEntity object.
        """
        ...

    @abstractmethod
    async def delete_one(
        self,
        role_id: int,
        session: Optional[Any] = None
    ) -> Optional[RoleEntity]:
        """
        Deletes an existing role.

        Parameters:
        - role_id: Role ID.
        - session: Session object.

        Returns:
        - RoleEntity: RoleEntity object.
        """
        ...

    @abstractmethod
    async def delete_many(
        self,
        role_ids: List[int],
        session: Optional[Any] = None
    ) -> List[RoleEntity]:
        """
        Deletes multiple existing roles.

        Parameters:
        - role_ids: List of role IDs.
        - session: Session object.

        Returns:
        - List: List of RoleEntity objects.
        """
        ...

File: cobalt/frontend/src/services/api/http/roles.ts
/*
 * Copyright (C) 2026 ArtoriasCode
 * Author: ArtoriasCode
 * Repository: https://github.com/ArtoriasCode/cobalt
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */

import type { IHttpClient, IHttpRolesApiService } from "@/contracts"
import type {
  RoleEntity,
  RolesPageEntity,
  RolesPageRequest,
  RoleCreateRequest,
  RoleUpdateRequest
} from "@/types"

/**
 * API service for roles endpoints.
 */
export class HttpRolesApiService implements IHttpRolesApiService {
  private readonly prefix = "/api/v1/roles"
  private readonly client: IHttpClient

  constructor(client: IHttpClient) {
    this.client = client
  }

  /**
   * Gets a paginated list of roles.
   *
   * Parameters:
   * - params: RolesPageRequest object.
   *
   * Returns:
   * - Promise<RolesPageEntity>: Paginated list of roles.
   */
  async getPage(params: RolesPageRequest): Promise<RolesPageEntity> {
    return this.client.get<RolesPageEntity>(`${this.prefix}/`, { params })
  }

  /**
   * Gets an existing role.
   *
   * Parameters:
   * - roleId: Role ID.
   *
   * Returns:
   * - Promise<RoleEntity>: Role entity.
   */
  async getOne(roleId: number): Promise<RoleEntity> {
    return this.client.get<RoleEntity>(`${this.prefix}/${roleId}`)
  }

  /**
   * Creates a new role.
   *
   * Parameters:
   * - data: RoleCreateRequest object.
   *
   * Returns:
   * - Promise<RoleEntity>: Created role entity.
   */
  async createOne(data: RoleCreateRequest): Promise<RoleEntity> {
    return this.client.post<RoleEntity>(`${this.prefix}/`, data)
  }

  /**
   * Updates an existing role.
   *
   * Parameters:
   * - roleId: Role ID.
   * - data: RoleUpdateRequest object.
   *
   * Returns:
   * - Promise<RoleEntity>: Updated role entity.
   */
  async updateOne(roleId: number, data: RoleUpdateRequest): Promise<RoleEntity> {
    return this.client.patch<RoleEntity>(`${this.prefix}/${roleId}`, data)
  }

  /**
   * Deletes an existing role.
   *
   * Parameters:
   * - roleId: Role ID.
   *
   * Returns:
   * - Promise<void>.
   */
  async deleteOne(roleId: number): Promise<void> {
    return this.client.delete<void>(`${this.prefix}/${roleId}`)
  }

  /**
   * Deletes multiple existing roles.
   *
   * Parameters:
   * - ids: List of role IDs.
   *
   * Returns:
   * - Promise<void>.
   */
  async deleteMany(ids: number[]): Promise<void> {
    return this.client.delete<void>(`${this.prefix}/batch`, { data: ids })
  }
}